src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DerSequenceReader.cs

```diff
@@ -0,0 +1,262 @@
+    internal class DerSequenceReader
+    {
+        private readonly byte[] _data;
+        private readonly int _end;
+        private int _position;
+
+        internal int ContentLength { get; private set; }
+
+        private DerSequenceReader(bool startAtPayload, byte[] data)
+        {
+            Debug.Assert(startAtPayload, "This overload is only for bypassing the sequence tag");
+            Debug.Assert(data != null, "Data is null");
+
+            _data = data;
+            _position = 0;
+            _end = data.Length;
+
+            ContentLength = data.Length;
+        }
+
+        internal DerSequenceReader(byte[] data)
+            : this(data, 0, data.Length)
+        {
+        }
+
+        internal DerSequenceReader(byte[] data, int offset, int length)
+        {
+            _data = data;
+            _end = offset + length;
+
+            Debug.Assert(data != null, "Data is null");
+            Debug.Assert(offset >= 0, "Offset is negative");
+            Debug.Assert(length > 2, "Length is too short");
+            Debug.Assert(data.Length >= offset + length, "Array is too short");
+
+            _position = offset;
+            EatTag(DerTag.Sequence);
+            ContentLength = EatLength();
+        }
+
+        internal static DerSequenceReader CreateForPayload(byte[] payload)
+        {
+            return new DerSequenceReader(true, payload);
+        }
+
+        internal bool HasData
+        {
+            get { return _position < _end; }
+        }
+
+        internal byte PeekTag()
+        {
+            return _data[_position];
+        }
+
+        internal void SkipValue()
+        {
+            EatTag((DerTag)PeekTag());
+            int contentLength = EatLength();
+            _position += contentLength;
+        }
+
+        internal int ReadInteger()
+        {
+            byte[] integerBytes = ReadIntegerBytes();
+
+            // integerBytes is currently Big-Endian, need to reverse it for
+            // Little-Endian to pass into BigInteger.
+            Array.Reverse(integerBytes);
+            BigInteger bigInt = new BigInteger(integerBytes);
+            return (int)bigInt;
+        }
+
+        internal byte[] ReadIntegerBytes()
+        {
+            EatTag(DerTag.Integer);
+
+            return ReadContentAsBytes();
+        }
+
+        internal byte[] ReadOctetString()
+        {
+            EatTag(DerTag.OctetString);
+
+            return ReadContentAsBytes();
+        }
+
+        internal string ReadOidAsString()
+        {
+            EatTag(DerTag.ObjectIdentifier);
+            int contentLength = EatLength();
+
+            // Each byte could cause 3 decimal characters to be written, plus a period. Over-allocate
+            // and avoid re-alloc.
+            StringBuilder builder = new StringBuilder(contentLength * 4);
+
+            byte firstByte = _data[_position];
+            byte first = (byte)(firstByte / 40);
+            byte second = (byte)(firstByte % 40);
```

Comment:
ISO/IEC 8825-1:2003 (BER/CER/DER encoding)

> ##### 8.19.4
  The numerical value of the first subidentifier is derived from the values of the first two object identifier  components in the object identifier value being encoded, using the formula:
    `(X*40)  +  Y`
    where X is the value of the first object identifier component and Y is the value of the second object identifier 
component. 
        <sub>NOTE – This packing of the first two object identifier components recognizes that only 
three values are allocated from the root node, and at most 39 subsequent values from nodes reached by X = 0 and X = 1</sub>

Bit partitioning would have been more computer-sciency :smile:.  I've taken the hint and amended the code:
```C#
            // The first byte is ((X * 40) + Y), where X is the first segment and Y the second.
            // ISO/IEC 8825-1:2003 section 8.19.4

            byte firstByte = _data[_position];
            byte first = (byte)(firstByte / 40);
            byte second = (byte)(firstByte % 40);
```